src/imitation/algorithms/preference_comparisons.py

```diff
@@ -448,6 +586,99 @@ def __call__(
+    ):
+        """Initialize the active selection fragmenter.
+
+        Args:
+            preference_predictor: a model that predicts the preference of frag A.
+            base_fragmenter: fragmenter instance to get
+                fragment pairs from trajectories
+            fragment_sample_factor: the factor of the number of
+                fragment pairs to sample from the base_fragmenter
+            uncertainty_on: the variable to calculate the variance on.
+                Can be logit|probability|label.
+            custom_logger: Where to log to; if None (default), creates a new logger.
+
+        Raises:
+            ValueError: `uncertainty_on` not in logit|probability|label.
+        """
+        super().__init__(custom_logger=custom_logger)
+        self.preference_predictor = preference_predictor
+        self.base_fragmenter = base_fragmenter
+        self.fragment_sample_factor = fragment_sample_factor
+        # self.preference_predictor = PreferencePredictor()
+        if not (uncertainty_on in ["logit", "probability", "label"]):
+            raise ValueError(f"{uncertainty_on} not supported.")
+        self.uncertainty_on = uncertainty_on
+
+    def __call__(
+        self,
+        trajectories: Sequence[TrajectoryWithRew],
+        fragment_length: int,
+        num_pairs: int,
+    ) -> Sequence[TrajectoryWithRewPair]:
+        # sample a large number (self.fragment_sample_factor*num_pairs)
+        # of fragments from all the trajectories
+        fragment_pairs = self.base_fragmenter(
+            trajectories,
+            fragment_length,
+            int(self.fragment_sample_factor * num_pairs),
+        )
+        var_estimates = []
+        for i, fragment in enumerate(fragment_pairs):
+            frag1, frag2 = fragment
+            trans1 = rollout.flatten_trajectories([frag1])
+            trans2 = rollout.flatten_trajectories([frag2])
+            if self.uncertainty_on == "logit":
+                with th.no_grad():
+                    # NOTE: Should we first sum the rewards
+                    # and then take the var (curr approach)
+                    # or first take the var and then sum the rewards?
+                    _, rew1_var = self._reward_moments(trans1)
+                    _, rew2_var = self._reward_moments(trans2)
+                var_estimates.append(rew1_var + rew2_var)
+            else:  # uncertainty_on is probability or label
+                with th.no_grad():
+                    rews1 = self.preference_predictor.rewards(trans1)
+                    rews2 = self.preference_predictor.rewards(trans2)
+                    probs = (
+                        self.preference_predictor.probability(rews1, rews2)
+                        .cpu()
+                        .numpy()
+                    )
+                if self.uncertainty_on == "probability":
+                    var_estimates.append(probs.var())
+                else:  # uncertainty_on is label
```

Comment:
This seems like it will just be capturing aleatoric uncertainty. Is that what we want?

Reply:
The PEBBLE paper's implementation seems to be calculating the variance over the probabilities given by the ensemble members. So that's why we have added 'probability' as one of the options for `uncertainty_on`. 
@AdamGleave Can you confirm if the implementation is correct?

Reply:
I think `abs(0.5 - probs)` would be aleatoric uncertainty and `probs.var()` is epistemic.